File: sources/askit_mobile/lib/api/answer_vote_rest_api.dart
import 'dart:convert';
import 'dart:developer';

import 'package:askit_mobile/api/rest_api_properties.dart';
import 'package:askit_mobile/entity/answer_vote.dart';
import 'package:http/http.dart' as http;

class AnswerVoteRestApi {
  static Future<AnswerVote> findByAnswerIdAndUserId(int answerId, int userId) async {
    var queryParameters = {'answerId': answerId.toString(), 'userId': userId.toString()};
    final path = 'api/answer-votes';
    final response = await http.get(Uri.http(RestApiProperties.url, path, queryParameters));

    if (response.statusCode != 200) {
      log(response.body);
      return null;
    }

    return AnswerVote.fromJson(jsonDecode(utf8.decode(response.bodyBytes)));
  }

  static Future<AnswerVote> save(AnswerVote object) async {
    final path = 'api/answer-votes';

    http.Response response = await http.post(Uri.http(RestApiProperties.url, path),
        headers: <String, String>{'Content-Type': 'application/json; charset=UTF-8'},
        body: jsonEncode(object.toJson()));

    if (response.statusCode != 200) {
      log(response.body);
      return null;
    }

    return AnswerVote.fromJson(jsonDecode(utf8.decode(response.bodyBytes)));
  }

  static Future<AnswerVote> update(AnswerVote object) async {
    final path = 'api/answer-votes';

    http.Response response = await http.put(Uri.http(RestApiProperties.url, path),
        headers: <String, String>{'Content-Type': 'application/json; charset=UTF-8'},
        body: jsonEncode(object.toJson()));

    if (response.statusCode != 200) {
      log(response.body);
      return null;
    }

    return AnswerVote.fromJson(jsonDecode(utf8.decode(response.bodyBytes)));
  }

  static Future<AnswerVote> deleteById(int id) async {
    final path = 'api/answer-votes/$id';

    http.Response response = await http.delete(Uri.http(RestApiProperties.url, path));

    if (response.statusCode != 200) {
      log(response.body);
      return null;
    }

    return AnswerVote.fromJson(jsonDecode(utf8.decode(response.bodyBytes)));
  }
}

File: sources/askit-mvc/src/main/java/com/project/askit/rest/api/AnswerVoteRestApi.java
package com.project.askit.rest.api;

import com.project.askit.entity.AnswerVote;
import com.project.askit.entity.Wrapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.core.env.Environment;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

@Component
public class AnswerVoteRestApi {

    private final Environment environment;

    Logger logger = LoggerFactory.getLogger(this.getClass());

    @Autowired
    public AnswerVoteRestApi(Environment environment) {
        this.environment = environment;
    }

    public AnswerVote findByAnswerIdAndUserId(int answerId,
                                              int userId) {
        AnswerVote result = null;
        try {
            String uri = environment.getProperty("rest.url.api") + "/answer-votes";
            String query = "?answerId=" + answerId + "&userId=" + userId;
            ResponseEntity<AnswerVote> response = new RestTemplate().exchange(
                    uri + query,
                    HttpMethod.GET,
                    null,
                    new ParameterizedTypeReference<>() {
                    });

            HttpStatus statusCode = response.getStatusCode();
            if (statusCode != HttpStatus.OK) {
                String errorMessage = uri + ": " + response.getBody();
                throw new Exception(errorMessage);
            }

            result = response.getBody();
        } catch (Exception e) {
            logger.error(e.getMessage());
            e.printStackTrace();
        }

        return result;
    }

    public Wrapper<AnswerVote> findAllByUserId(Integer page,
                                               Integer size,
                                               String sort,
                                               String order,
                                               int userId) {
        Wrapper<AnswerVote> result = null;
        try {
            String query = "?page=" + (page == null ? "" : page) +
                    "&size=" + (size == null ? "" : size) +
                    "&sort=" + (sort == null ? "" : sort) +
                    "&order=" + (order == null ? "" : order);
            String uri = environment.getProperty("rest.url.api") + "/answer-votes/user/" + userId + query;
            ResponseEntity<Wrapper<AnswerVote>> response = new RestTemplate().exchange(
                    uri,
                    HttpMethod.GET,
                    null,
                    new ParameterizedTypeReference<>() {
                    });

            HttpStatus statusCode = response.getStatusCode();
            if (statusCode != HttpStatus.OK) {
                String errorMessage = uri + ": " + response.getBody();
                throw new Exception(errorMessage);
            }

            result = response.getBody();
        } catch (Exception e) {
            logger.error(e.getMessage());
            e.printStackTrace();
        }

        return result;
    }

    public AnswerVote save(AnswerVote object) {
        AnswerVote result = null;

        try {
            String uri = environment.getProperty("rest.url.api") + "/answer-votes";
            HttpEntity<AnswerVote> requestEntity = new HttpEntity<>(object, null);

            ResponseEntity<AnswerVote> response = new RestTemplate().exchange(
                    uri,
                    HttpMethod.POST,
                    requestEntity,
                    new ParameterizedTypeReference<>() {
                    });

            HttpStatus statusCode = response.getStatusCode();
            if (statusCode != HttpStatus.OK) {
                String errorMessage = uri + ": " + response.getBody();
                throw new Exception(errorMessage);
            }

            result = response.getBody();
        } catch (Exception e) {
            logger.error(e.getMessage());
            e.printStackTrace();
        }

        return result;
    }

    public AnswerVote update(AnswerVote object) {
        AnswerVote result = null;

        try {
            String uri = environment.getProperty("rest.url.api") + "/answer-votes";
            HttpEntity<AnswerVote> requestEntity = new HttpEntity<>(object, null);

            ResponseEntity<AnswerVote> response = new RestTemplate().exchange(
                    uri,
                    HttpMethod.PUT,
                    requestEntity,
                    new ParameterizedTypeReference<>() {
                    });

            HttpStatus statusCode = response.getStatusCode();
            if (statusCode != HttpStatus.OK) {
                String errorMessage = uri + ": " + response.getBody();
                throw new Exception(errorMessage);
            }

            result = response.getBody();
        } catch (Exception e) {
            logger.error(e.getMessage());
            e.printStackTrace();
        }

        return result;
    }

    public AnswerVote deleteById(Integer id) {
        AnswerVote result = null;

        try {
            String uri = environment.getProperty("rest.url.api") + "/answer-votes/" + id;
            ResponseEntity<AnswerVote> response = new RestTemplate().exchange(
                    uri,
                    HttpMethod.DELETE,
                    null,
                    new ParameterizedTypeReference<>() {
                    });

            HttpStatus statusCode = response.getStatusCode();
            if (statusCode != HttpStatus.OK) {
                String errorMessage = uri + ": " + response.getBody();
                throw new Exception(errorMessage);
            }

            result = response.getBody();
        } catch (Exception e) {
            logger.error(e.getMessage());
            e.printStackTrace();
        }

        return result;
    }
}
